polytracker/include/taintdag/fnmapping.h

```diff
@@ -9,74 +9,40 @@
 #pragma once
 
 #include <cstdint>
-#include <mutex>
 #include <optional>
 #include <string_view>
-#include <unordered_map>
 
-// TODO(hbrodin): Currently using this as a workaround to be able to compile
-// this file as part of FunctionTracingPass.
-#if __cpp_concepts
-#if __cpp_lib_concepts
 #include "taintdag/outputfile.h"
 #include "taintdag/section.h"
 #include "taintdag/string_table.h"
-#endif
-#endif
 
 namespace taintdag {
 
-class FnMapping {
+struct Function {
 public:
   using offset_t = uint32_t;
   using length_t = uint32_t;
+  offset_t name_offset;
+  length_t name_len;
+};
 
-  struct header_t {
-    offset_t name_offset;
-    length_t name_len;
-  };
-
+class Functions : public FixedSizeAlloc<Function> {
+public:
   using index_t = uint16_t;
 
-  FnMapping(char *begin, char *end);
-  std::optional<index_t> add_mapping(std::string_view name);
-  size_t get_mapping_count();
-
-private:
-  // Map markers
-  char *map_begin{nullptr};
-  char *map_end{nullptr};
-  char *headers_end{nullptr};
-  char *names_begin{nullptr};
-  // Map mutex
-  std::mutex memory_m;
-  // Helpers
-  std::optional<offset_t> write_name(std::string_view name);
-  std::optional<offset_t> write_header(header_t header);
-  // Cache
-  std::unordered_map<std::string_view, index_t> mappings;
-};
-
-// TODO (hbrodin): Unsure what goes in the FunctionEntry atm.
-struct FunctionEntry {};
+  static constexpr uint8_t tag{6};
+  static constexpr size_t allocation_size{std::numeric_limits<index_t>::max() *
```

Comment:
I think you're right. I've used [`Sources`](https://github.com/trailofbits/polytracker/blob/new_fn_mapping/polytracker/include/taintdag/taint_source.h#L59-L60) as reference. Does this mean that the code in `Sources` should also be modified? Or is there a reason why there's not a `+1` there?

Reply:
Ohh... Yeah, I believe there should be a +1 there as well. I believe nothing bad will happen, we just won't be able to store the last item. There will be a limit check in the `construct`-method to prevent out of bounds writes.